eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs

```diff
@@ -450,76 +456,90 @@ getRewardAccountAmount rewardAcount = do
   kpSpending <- lookupKeyPair =<< freshKeyHash
   kpStaking <- lookupKeyPair =<< freshKeyHash
+  (rootTxIn, rootTxOut) <- lookupImpRootTxOut
   let
     deposits = certsTotalDepositsTxBody pp certState (tx ^. bodyTxL)
     refunds = certsTotalRefundsTxBody pp certState (tx ^. bodyTxL)
     outputsTotalCoin = sumAllCoin $ tx ^. bodyTxL . outputsTxBodyL
-    remainingCoin = impRootTxCoin <-> (outputsTotalCoin <+> deposits <-> refunds)
+    remainingCoin = (rootTxOut ^. coinTxOutL) <-> (outputsTotalCoin <+> deposits <-> refunds)
     remainingTxOut =
       mkBasicTxOut @era
         (mkAddr (kpSpending, kpStaking))
         (inject remainingCoin)
-  impRootTxCoinL .= remainingCoin
   let
-    balancedTx =
-      setMinFeeTx pp tx
-        & bodyTxL . inputsTxBodyL %~ Set.insert (TxIn impRootTxId $ TxIx 0)
+    txNoWits =
+      tx
+        & bodyTxL . inputsTxBodyL %~ Set.insert rootTxIn
         & bodyTxL . outputsTxBodyL %~ (remainingTxOut :<|)
-    txId = TxId . hashAnnotated $ balancedTx ^. bodyTxL
-  impRootTxIdL .= txId
-  wits <- mkTxWits (balancedTx ^. bodyTxL)
-  pure $ balancedTx & witsTxL .~ wits
+    nativeScriptKeyWitsCount = 0 -- TODO figure out wits for native scripts
+    outsWithoutRemaining = tx ^. bodyTxL . outputsTxBodyL
+    fee = calcMinFeeTx utxo pp txNoWits nativeScriptKeyWitsCount
+    remainderAvailable = remainingTxOut ^. coinTxOutL <-> fee
+    remainingTxOut' = remainingTxOut & coinTxOutL .~ remainderAvailable
+    -- If the remainder is sufficently big we add it to outputs, otherwise we add the
+    -- extraneous coin to the fee and discard the remainder TxOut
+    txBalanced
+      | remainderAvailable >= getMinCoinTxOut pp remainingTxOut' =
+          txNoWits
+            & bodyTxL . outputsTxBodyL .~ (remainingTxOut' :<| outsWithoutRemaining)
+            & bodyTxL . feeTxBodyL .~ fee
+      | otherwise =
+          txNoWits
+            & bodyTxL . outputsTxBodyL .~ outsWithoutRemaining
+            & bodyTxL . feeTxBodyL .~ (fee <> remainingTxOut ^. coinTxOutL)
+  wits <- mkTxWits (txBalanced ^. bodyTxL)
+  let txWithWits = txBalanced & witsTxL .~ wits
+      Coin feeUsed = txWithWits ^. bodyTxL . feeTxBodyL
+      Coin feeMin = getMinFeeTx pp txWithWits
+  when (feeUsed /= feeMin) $ do
+    logEntry $
+      "Estimated fee " <> show feeUsed <> " while required fee is " <> show feeMin
+  pure txWithWits
+
+submitTx_ :: (HasCallStack, ShelleyEraImp era) => Tx era -> ImpTestM era ()
+submitTx_ = void . submitTx
+
+submitTx :: (HasCallStack, ShelleyEraImp era) => Tx era -> ImpTestM era (Tx era)
+submitTx tx = trySubmitTx tx >>= expectRightDeepExpr
 
-submitTx_ ::
-  forall era.
+trySubmitTx ::
   ShelleyEraImp era =>
   Tx era ->
-  ImpTestM
-    era
-    ( Either
-        [PredicateFailure (EraRule "LEDGER" era)]
-        (State (EraRule "LEDGER" era), Tx era)
-    )
-submitTx_ tx = do
+  ImpTestM era (Either [PredicateFailure (EraRule "LEDGER" era)] (Tx era))
```

Comment:
I think I'd prefer returning the `TxId` rather than the `Tx` in all of the ImpTest functions that deal with submitting a transaction. Returning a `Tx` would also work since you can access the `TxId` from it as well, but it's a good idea to agree on which one we're going to use and then use that across all `ImpTest` modules. 

The way I'd like to think about it is that `Tx` is used for transactions that are not yet submitted to the chain and `TxId` represents a transaction on the chain.

Reply:
I totally hear you. However, the problem is that transaction that was supplied to the function is not the transaction that gets submitted. We modify it by fixing up the fee and by adding witnesses to it. Without returning the modified transaction there is no way to get access to what has been submitted, which can be useful for some tests.

That was the main reason why I switched from the `TxId` to `Tx`.